lenskit-solver/src/main/java/org/lenskit/featurizer/FeatureExtractor.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+package org.lenskit.featurizer;
+
+import org.lenskit.data.entities.Entity;
+import org.lenskit.space.IndexSpace;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The general interface for feature extractors.
+ *
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+public interface FeatureExtractor extends Serializable {
```

Comment:
Consider dropping the `Map` from `extract`, instead just returning a `List<Feature>`. If we do this, then we can add a `getFeatureName()` method to `FeatureExtractor` to tell the learning model / featurizer what this feature group name is. Alternatively, the learning model or featurizer could associate names with feature extractors. Either way, it will make the feature extractor interface simpler and easier to understand.